cpp/src/arrow/status.h

```diff
@@ -157,8 +157,10 @@ class ARROW_EXPORT [[nodiscard]] Status : public util::EqualityComparable<Status
   inline bool Equals(const Status& s) const;
 
   // AND the statuses.
-  inline Status operator&(const Status& s) const noexcept;
-  inline Status operator&(Status&& s) const noexcept;
+  inline Status operator&(const Status& s) const& noexcept;
+  inline Status operator&(const Status& s) && noexcept;
+  inline Status operator&(Status&& s) const& noexcept;
+  inline Status operator&(Status&& s) && noexcept;
```

Comment:
Looks to me "pass by value" is simpler. To reduce half of the functions. There might be overhead of an additional move, should be trivial.
```
Status operator&(Status s) const&;
Status operator&(Status s) &&;
```

Reply:
I do not understand what you mean by the "simpler" that you mentioned. 

It should be like this in c++23 or later:

```c++
template <typename Self, typename T, std::enable_if_t<std::is_same_v<std::remove_cvref_t<T>, Status>, int> = 0>
Status operator&(this Self&& self, T&& other) {
     if (ok()) return std::forward<T>(other);
    return std::forward<Self>(self);
}
```

(hand-written, you can check it by compiler and lang-server)